mitmproxy/dns.py

```diff
@@ -0,0 +1,869 @@
+            answers=[],
+            authorities=[],
+            additionals=[],
+        )
+
+    def succeed(self, answers: List[ResourceRecord]) -> Message:
+        return Message(
+            timestamp=time.time(),
+            id=self.id,
+            query=False,
+            op_code=self.op_code,
+            authoritative_answer=False,
+            truncation=False,
+            recursion_desired=self.recursion_desired,
+            recursion_available=True,
+            reserved=0,
+            response_code=ResponseCode.NOERROR,
+            questions=self.questions,
+            answers=answers,
+            authorities=[],
+            additionals=[],
+        )
+
+    def resolve(self) -> Message:
+        """Resolves the message and return the result in form of a response message."""
+        try:
+            if not self.query:
+                raise ResolveError(ResponseCode.REFUSED)  # we cannot resolve an answer
+            if self.op_code is not OpCode.QUERY:
+                raise ResolveError(ResponseCode.NOTIMP)  # inverse queries and others are not supported
+            rrs = [rr for rrs in map(lambda q: q.resolve(), self.questions) for rr in rrs]
+        except ResolveError as e:
+            return self.fail(e.response_code)
+        else:
+            return self.succeed(rrs)
+
+    @classmethod
+    def unpack(cls, buffer: bytes) -> Message:
+        """Converts the entire given buffer into a DNS message."""
+        length, msg = cls.unpack_from(buffer, 0)
+        if length != len(buffer):
+            raise struct.error(f"unpack requires a buffer of {length} bytes")
+        return msg
+
+    @classmethod
+    def unpack_from(cls, buffer: Union[bytes, bytearray], offset: int) -> Tuple[int, Message]:
+        """Converts the buffer from a given offset into a DNS message and also returns its length."""
+        id, flags, len_questions, len_answers, len_authorities, len_additionals = Message.HEADER.unpack_from(buffer, offset)
+        try:
+            msg = Message(
+                timestamp=time.time(),
+                id=id,
+                query=(flags & (1 << 15)) == 0,
+                op_code = OpCode((flags >> 11) & 0b1111),
+                authoritative_answer=(flags & (1 << 10)) != 0,
+                truncation = (flags & (1 << 9)) != 0,
+                recursion_desired = (flags & (1 << 8)) != 0,
+                recursion_available = (flags & (1 << 7)) != 0,
+                reserved = (flags >> 4) & 0b111,
+                response_code = ResponseCode(flags & 0b1111),
+                questions=[],
+                answers=[],
+                authorities=[],
+                additionals=[],
+            )
+        except ValueError as e:
+            raise struct.error(str(e))
+        offset += Message.HEADER.size
+        cached_names: Dict[int, Optional[Tuple[str, int]]] = dict()
+
+        def unpack_domain_name() -> str:
+            nonlocal buffer, offset
```

Comment:
Somewhat related: For HTTP/1, we have tucked all packing/unpacking into https://github.com/mitmproxy/mitmproxy/tree/main/mitmproxy/net/http/http1. It may make sense to be consistent here for UDP and not make packing/unpacking part of the public API for now.

Reply:
Agreed, moved in f535045e69207402f224f8f28c355caec84040e6 to `mitmproxy.net.dns` and simplified common code. The difference between the two implementations is DNS compression support. Also renamed the functions to make that clearer. Introduced `domain_name.cache` for (un)compressed domain names, which can be used transparently by invoking code, instead of relying that it's a `dict`.